File: src/main/java/lsifjava/ProjectIndexer.java
package lsifjava;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

public class ProjectIndexer {
    public int numFiles;
    public int numDefinitions;

    private final Arguments arguments;
    private final Emitter emitter;

    public ProjectIndexer(Arguments arguments, Emitter emitter) {
        this.arguments = arguments;
        this.emitter = emitter;
    }

    public void index() throws IOException {
        emitter.emitVertex("metaData", Map.of(
            "version", "0.4.0",
            "positionEncoding", "utf-16",
            "projectRoot", String.format("file://%s", Paths.get(arguments.projectRoot).toFile().getCanonicalPath()),
            "toolInfo", Map.of("name", "lsif-java", "version", "0.1.0")
        ));

        var projectId = emitter.emitVertex("project", Map.of(
            "kind", "java"
        ));

        var indexers = new HashMap<Path, DocumentIndexer>();
        var collector = new FileCollector(projectId, arguments, emitter, indexers);

        try(GradleInterface gradleInterface = new GradleInterface(arguments.projectRoot)) {
            for (String it : gradleInterface.sourceDirectories()) {
                Files.walkFileTree(Paths.get(it), collector);
            }
        }

        var fileManager = new SourceFileManager(indexers.keySet());

        for(var indexer : indexers.values()) {
            indexer.preIndex(fileManager);
        }

        for(var indexer : indexers.values()) {
            indexer.index();
        }

        /* if(arguments.verbose)
            System.out.println("\nEMITTING DEFINITIONS\n-------------------------------------------------");

        for(var indexer : indexers.values()) {
            indexer.visitDefinitions();
        }

        if(arguments.verbose)
            System.out.println("\nEMITTING REFERENCES\n-------------------------------------------------");

        for(var indexer : indexers.values()) {
            indexer.visitReferences();
        }
 */
        for(var indexer : indexers.values()) {
            indexer.postIndex();
        }

        for(var indexer : indexers.values()) {
            numFiles++;
            numDefinitions += indexer.numDefinitions();
        }
    }
}

File: src/main/kotlin/lsifjava/ProjectIndexer.kt
package lsifjava

internal class IndexingException(path: String, cause: Throwable): Throwable("failed to index $path", cause)

class ProjectIndexer(private val arguments: Arguments, private val emitter: Emitter) {
    var numFiles = 0
    var numDefinitions = 0
    var numJavacErrors = 0
    
    fun index() {
        emitter.emitVertex("metaData", mapOf(
            "version" to "0.4.0",
            "positionEncoding" to "utf-16",
            "projectRoot" to "file://${arguments.projectRoot}",
            "toolInfo" to mapOf("name" to "lsif-java", "version" to "0.7.0")
        ))

        val projectId = emitter.emitVertex("project", mapOf(
            "kind" to "java"
        ))

        createJavacDiagnosticListener(arguments.javacOutWriter).use { javacDiagListener ->
            val indexers = GradleInterface(arguments.projectRoot).use {
                buildIndexerMap(it, emitter, arguments.verbose, javacDiagListener)
            }

            val fileManager = SourceFileManager(indexers.keys)

            // can we do this earlier???
            for (indexer in indexers.values) {
                indexer.fileManager = fileManager
            }

            for (indexer in indexers.values) {
                try {
                    indexer.index()
                } catch(e: IndexingException) {
                    println(e)
                }
            }

            for (indexer in indexers.values) {
                indexer.postIndex(projectId)
            }

            for (indexer in indexers.values) {
                numFiles++
                numDefinitions += indexer.numDefinitions
            }

            numJavacErrors = javacDiagListener.count
        }
    }
}